book-content/chapters/16-the-utils-folder.md

```diff
@@ -0,0 +1,1527 @@
+
+In this case, the implementation signature will take in a parameter called `queryOrCriteria` that can be either a `string` or an object with the specified properties. Inside the function, we'll check the type of `queryOrCriteria` and perform the appropriate search logic based on the provided arguments:
+
+```typescript
+function searchMusic(artist: string, genre: string, year: number): void;
+function searchMusic(criteria: {
+  artist: string;
+  genre: string;
+  year: number;
+}): void;
+function searchMusic(
+  artistOrCriteria: string | { artist: string; genre: string; year: number },
+  genre?: string,
+  year?: number,
+): void {
+  if (typeof artistOrCriteria === "string") {
+    // Search with separate arguments
+    search(artistOrCriteria, genre, year);
+  } else {
+    // Search with object
+    search(
+      artistOrCriteria.artist,
+      artistOrCriteria.genre,
+      artistOrCriteria.year,
+    );
+  }
+}
+```
+
+Now we can call the `searchMusic` function with the different arguments defined in the overloads:
+
+```typescript
+searchMusic("King Gizzard and the Lizard Wizard", "Psychedelic Rock", 2021);
+searchMusic({
+  artist: "Tame Impala",
+  genre: "Psychedelic Rock",
+  year: 2015,
+});
+```
+
+However, TypeScript will warn us if we attempt to pass in an argument that doesn't match any of the defined overloads:
+
+```ts twoslash
+// @errors: 2575
+function searchMusic(artist: string, genre: string, year: number): void;
+function searchMusic(criteria: {
+  artist: string;
+  genre: string;
+  year: number;
+}): void;
+function searchMusic(
+  artistOrCriteria: string | { artist: string; genre: string; year: number },
+  genre?: string,
+  year?: number,
+): void {}
+// ---cut---
+searchMusic(
+  {
+    artist: "Tame Impala",
+    genre: "Psychedelic Rock",
+    year: 2015,
+  },
+  "Psychedelic Rock",
+);
+```
+
+This error shows us that we're trying to call `searchMusic` with two arguments, but the overloads only expect one or three arguments.
+
+### Function Overloads vs Unions
+
+Function overloads can be useful when you have multiple call signatures spread over different sets of arguments. In the example above, we can either call the function with one argument, or three.
+
+When you have the same number of arguments but different types, you should use a union type instead of function overloads. For example, if you want to allow the user to search by either artist name or criteria object, you could use a union type:
+
+```typescript
+function searchMusic(
+  query: string | { artist: string; genre: string; year: number },
+): void {
+  if (typeof query === "string") {
+    // Search by artist
+    searchByArtist(query);
+  } else {
+    // Search by all
+    search(query.artist, query.genre, query.year);
+  }
+}
+```
+
+This uses far fewer lines of code than defining two overloads and an implementation.
```

Comment:
pseudo-overloads are also sometimes nice:
```ts
function test(...args: [string, number] | [number, string]) {}
```

but, unfortunately, they don't allow us to link parameter types with return types anyhow